airbyte-ci/connectors/pipelines/pipelines/cli/auto_update.py

```diff
@@ -0,0 +1,122 @@
+# HELPERS
+
+import importlib
+import logging
+import os
+import sys
+
+import asyncclick as click
+import requests
+from pipelines import main_logger
+from pipelines.consts import LOCAL_PIPELINE_PACKAGE_PATH
+
+__installed_version__ = importlib.metadata.version("pipelines")
+
+BINARY_UPGRADE_COMMAND = "make tools.airbyte-ci.install"
+DEV_UPGRADE_COMMAND = "make tools.airbyte-ci-dev.install"
+
+
+def _get_os_name():
+    """
+    Returns 'ubuntu' if the system is Linux or 'macos' if the system is macOS.
+    """
+    OS = os.uname().sysname
+    if OS == "Linux":
+        return "ubuntu"
+    elif OS == "Darwin":
+        return "macos"
+    else:
+        # Default to macos as this is just a check, if they are not supported they will find
+        # out at install time.
+        return "macos"
+
+
+def _is_upgrade_available(version: str, is_dev: bool) -> bool:
+    """
+    Check if an upgrade is available for the given version.
+    """
+
+    # Given that they can install from source, we don't need to check for upgrades
+    if is_dev:
+        return True
+
+    # Get RELEASE_URL from the environment if it exists, otherwise use the default
+    # "https://connectors.airbyte.com/files/airbyte-ci/releases"
+    release_url = os.getenv("RELEASE_URL", "https://connectors.airbyte.com/files/airbyte-ci/releases")
+    os_name = _get_os_name()
+    url = f"{release_url}/{os_name}/{version}/airbyte-ci"
+
+    # Just check if the URL exists, but dont download it
+    return requests.head(url).status_code == 200
+
+
+def _get_latest_version() -> str:
+    """
+    Get the version of the latest release, which is just in the pyproject.toml file of the pipelines package
+    as this is an internal tool, we don't need to check for the latest version on PyPI
+    """
+    path_to_pyproject_toml = LOCAL_PIPELINE_PACKAGE_PATH + "pyproject.toml"
+    with open(path_to_pyproject_toml, "r") as f:
+        for line in f.readlines():
+            if "version" in line:
+                return line.split("=")[1].strip().replace('"', "")
+    raise Exception("Could not find version in pyproject.toml. Please ensure you are running from the root of the airbyte repo.")
```

Comment:
I'd prefer our release process to determine what is the latest version by writing a file like `version.txt` under `airbyte-ci/releases/latest/version.txt` in the GCS bucket.
It will decouple the bin upgrade from the repo and make sure the `pyproject.toml` is actually available on GCS first...

Reply:
This is a fair callout but what do we do when the user is running an old version of airbyte? Im thinking like a feature branch they havent rebased. Where airbyte-ci is set to v1 but on latest its v2.

In this case we wouldnt want them to install latest as our airbyte-ci features often depend of the state of the airbyte code base it was released in

Reply:
But `make tools.airbyte-ci.install` always install latest and does not depend on the repo state right?

> In this case we wouldnt want them to install latest as our airbyte-ci features often depend of the state of the airbyte code base it was released in

If a new airbyte-ci version is incompatible with the current code base I think we should denote this with a major release. As users are prompted on auto-update they will decide to install or not a new version, and the other command in this PR make it easy to roll back to a different version if needed

Reply:
> But make tools.airbyte-ci.install always install latest and does not depend on the repo state right?
Thats actually only partially right. We recommend regular users use `make tools.airbyte-ci.install` however there is a version argument `make tools.airbyte-ci.install VERSION=1.2.3` which is available and is what auto_upgrade uses.

> If a new airbyte-ci version is incompatible with the current code base I think we should denote this with a major release.
Im not sure I 100% agree. I see the spirit of this but im not so sure it matches the reality of some of our OSS customers.

If a OSS user deploys airbyte in January and we release a major version of `airbyte-ci` in November. We want to avoid these scenarios
1. They are forced to use `airbyte-ci` latest which is now incompatible with the airbyte they deployed
2. We are forced to build in backwards compatibility into subsequent versions of `airbyte-ci`

I think the only way we can avoid that is to have somewhere that denotes this version of the airbyte code base can run with this version of airbyte-ci.

Reply:
👍 Right. This approach seems safer indeed, thank you for the example.

Reply:
Thanks for the good question!